funnel/assets/js/account_form.js

```diff
@@ -0,0 +1,187 @@
+import 'htmx.org';
+import Form from './utils/formhelper';
+
+window.Hasgeek.Accountform = ({
+  formId,
+  usernameField,
+  passwordField,
+  passwordForm,
+  accountUsernameUrl,
+  passwordCheckUrl,
+}) => {
+  $(formId)
+    .parsley()
+    .subscribe('parsley:field:validated', () => {
+      if ($(formId).parsley().isValid())
+        $(formId).addClass('parsley-valid').removeClass('parsley-invalid');
+      else $(formId).addClass('parsley-invalid').removeClass('parsley-valid');
+    });
+
+  let typingTimer;
+  let typingTimerUsername;
+  const typingWaitInterval = 100;
+  let waitingForResponse = false;
+  let waitingForUsernameResponse = false;
+
+  async function checkUsernameAvailability(field) {
+    if (!waitingForUsernameResponse && $(field).val()) {
+      waitingForUsernameResponse = true;
+      const response = await fetch(accountUsernameUrl, {
+        method: 'POST',
+        headers: {
+          Accept: 'application/json',
+          'Content-Type': 'application/x-www-form-urlencoded',
+          'X-Requested-With': 'XMLHttpRequest',
+        },
+        body: new URLSearchParams({
+          username: $(field).val(),
+          csrf_token: $('meta[name="csrf-token"]').attr('content'),
+        }).toString(),
+      }).catch(() => {
+        window.toastr.error(window.Hasgeek.Config.errorMsg.networkError);
+      });
+      if (response && response.ok) {
+        const remoteData = await response.json();
+        if (remoteData) {
+          waitingForUsernameResponse = false;
+          const fieldWrapper = $(field).closest('.mui-form__controls');
+          if (remoteData.status === 'error') {
+            $(field).closest('.mui-form__fields').addClass('has-error');
+            if ($(fieldWrapper).find('p.mui-form__error').length) {
+              $(fieldWrapper)
+                .find('p.mui-form__error')
+                .text(remoteData.error_description);
+            } else {
+              const errorTxt = $('<p class="mui-form__error"></p>').text(
+                remoteData.error_description
+              );
+              $(errorTxt).insertBefore($(fieldWrapper).find('.mui-form__helptext'));
+            }
+          } else {
+            $(field).closest('.mui-form__fields').removeClass('has-error');
+            $(fieldWrapper).find('p.mui-form__error').remove();
+          }
+        }
+      } else {
+        waitingForUsernameResponse = false;
+        Form.getFetchError(response);
+      }
+    }
+  }
+
+  async function checkPasswordStrength(field) {
+    // FIXME: This function is getting called multiple times for same candidate
```

Comment:
Sounds like endpoint needs to set cache headers.

Reply:
Debouncing is present, will that suffice?
https://github.com/hasgeek/funnel/blob/main/funnel/templates/account_formlayout.html.jinja2#L84